phlsys_fs: Add 'delete_file_if_exists' and 'write_text_file_atomic'

cs-shadow opened the pull request:
Add two functions which will be useful in various places -
'delete_file_if_exists' which silently deletes a file, without raising
an error if it does not exists and 'write_text_file_atomic' which writes
text to a file as one atomic operation (using a temporary file) and also
takes care of creating intermediate directories and the file.

Test plan:
$ ./precommit.sh

aevri commented on py/phl/phlsys_fs.py:
```diff
@@ -338,6 +341,49 @@ def write_text_file(path, text):
+    dir_path = os.path.dirname(path)
+    if dir_path:
+        ensure_dir(dir_path)
+    tmppath = path + "~"
```
This same file path would be picked for contending atomic writers to the same file, so this would still have a race in a multiple writers scenario. It would be safe from a single reader+writer race though.

We could use `tempfile.mkstemp()` to safely create a unique file, then write our content, then set the permissions according to umask, and finally move it into place. I think that would be safe in the multiple writers scenario.

cs-shadow replied:
I see, I didn't consider the multiple writer situation while writing it but have changed it now.

cs-shadow commented:
@aevri This is ready for review now. As per our discussion, I thought it'll be best to use `tempfile.mkstemp()` function as it is without trying to `fchmod()` it based on `umask()` since that may have very serious repercussions if its interrupted in between as python does not provides a method to read umask value in an atomic fashion. I couldn't find an elegant way to do that as well. I've added documentation to the function's docblock mentioning the file's permissions.

Also, the lacks of tests in file read/write functions was bothering me so I added doc tests for those as well. Although the functions are trivial but still it would be nice to cover all bases.

aevri commented on py/phl/phlsys_fs.py:
```diff
@@ -338,6 +369,61 @@ def write_text_file(path, text):
+    Usage example:
+
+        >>> with chtmpdir_context():
+        ...     write_text_file('testfile', 'Hello')
```
This should probably be `write_text_file_atomic` :)

cs-shadow replied:
Yes, my bad - too much copy-paste is not good.

aevri commented on py/phl/phlsys_fs.py:
```diff
@@ -338,6 +369,61 @@ def write_text_file(path, text):
+
+def write_text_file_atomic(path, text):
+    """Atomically writes 'text' to the file at 'path' using temporary file.
+
```
It might be good to explain the atomic part here a bit too, i.e. "if multiple writers, reader will always see one consistent version". We could also say that this is achieved by "overwrite-by-rename". This seems to be a good answer for inspiration: http://unix.stackexchange.com/a/45812

aevri commented on py/phl/phlsys_fs.py:
```diff
@@ -338,6 +369,61 @@ def write_text_file(path, text):
+    tmpfile, tmppath = tempfile.mkstemp(text=True)
+    tmpfile.write(text)
+    tmpfile.close()
+    os.rename(tmppath, path)
```
Implementation looks good, thanks!

cs-shadow commented:
Updated the PR to address your concerns.

aevri commented:
Landed at 0264f45, thanks!
